source/extensions/priv/server/ntds_decrypt.c

```diff
@@ -0,0 +1,194 @@
+	if (decryptFromRID(encodedHash, rid, decodedHash) != 0){
+		return FALSE;
+	}
+	return TRUE;
+}
+
+/*!
+* @brief Splits up an encrypted LM or NT hash history and decrypts each stored hash
+* @param encHashHistory Pointer to the byte array containing the hash history record
+* @param sizeHistory The size of the history record
+* @param pekDecrypted Pointer to a decryptedPEK structure that holds our decrypted PEK
+* @param rid DWORD representing the Relative ID(RID) of the account
+* @param accountHistory Pointer to a string wherewe store all the decrypted historical hashes
+* @param historyCount Pointer to n integer where we store a count of the historical hashes
+* @returns Indication of sucess or failure.
+*/
+BOOL decrypt_hash_history(LPBYTE encHashHistory, size_t sizeHistory, struct decryptedPEK *pekDecrypted, DWORD rid, char *accountHistory, int *historyCount){
+	BOOL cryptOK = FALSE;
+	int sizeHistoryData = (int)sizeHistory - 24;
+	int numHashes = (sizeHistoryData / HASH_LENGTH_BYTES);
+	memcpy(historyCount, &numHashes, sizeof(historyCount));
+	LPBYTE encHistoryData = (LPBYTE)calloc(1,sizeHistoryData);
+	LPBYTE decHistoryData = (LPBYTE)calloc(1,(sizeHistoryData * 2));
+	memcpy(encHistoryData, encHashHistory + 24, sizeHistoryData);
+	cryptOK = decrypt_rc4(pekDecrypted->pekKey, encHashHistory + 8, encHistoryData, 1, sizeHistoryData);
+	if (!cryptOK){
+		free(encHistoryData);
+		free(decHistoryData);
+		return FALSE;
+	}
+	LPBYTE historicalHash = encHistoryData;
+	LPBYTE writeMarker = decHistoryData;
+	for (int i = 0; i < numHashes; i++){
+		BYTE decHash[HASH_LENGTH_BYTES];
+		char hashString[NULL_TERIMNATED_HASH_STRING_LENGTH];
+		cryptOK = decrypt_hash_from_rid(historicalHash, &rid, decHash);
+		if (!cryptOK){
+			return FALSE;
+		}
+		bytes_to_string(decHash, HASH_LENGTH_BYTES, hashString);
+		strncpy(writeMarker, hashString, NULL_TERIMNATED_HASH_STRING_LENGTH);
+		historicalHash = historicalHash + HASH_LENGTH_BYTES;
+		writeMarker = writeMarker + NULL_TERIMNATED_HASH_STRING_LENGTH;
+	}
+	memcpy(accountHistory, decHistoryData, (numHashes * NULL_TERIMNATED_HASH_STRING_LENGTH));
+	free(encHistoryData);
+	free(decHistoryData);
+	return TRUE;
+}
+
+/*!
+* @brief Decrypts the Password Encryption Key(PEK)
+* @param sysKey Pointer to the string holding the SYSKEY
+* @param pekEncrypted Pointer to an encryptedPEK struct containing our PEK to be decrypted
+* @param pekDecrypted Pointer to the decryptedPEK struct where we will store our decrypted PEK
+* @returns Indication of sucess or failure.
+*/
+BOOL decrypt_PEK(unsigned char *sysKey, struct encryptedPEK *pekEncrypted, struct decryptedPEK *pekDecrypted){
+	BOOL cryptOK = FALSE;
+	BYTE pekData[52] = { 0 };
+	DWORD pekLength = 52;
+	memcpy(&pekData, &pekEncrypted->pekData, pekLength);
+
+	cryptOK = decrypt_rc4(sysKey, pekEncrypted->keyMaterial, pekData, 1000, pekLength);
+	if (!cryptOK){
+		return FALSE;
+	}
+	memcpy(pekDecrypted, &pekData, pekLength);

```

Comment:
I feel like it should copy sizeof(struct decryptedPEK) to avoid doubts about the length copied.